File: src/parser_utils/output/__tests__/js_layout_generator/assignment_precedence_test.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open Ast_builder
open Layout_test_utils
module S = Ast_builder.Statements
module E = Ast_builder.Expressions
module L = Layout_builder

let test ctxt =
  let rhs = E.sequence [E.identifier "y"; E.identifier "z"] in
  let ast = E.assignment (Patterns.identifier "x") rhs in
  assert_layout_of_expression
    ~ctxt
    L.(
      loc
        (fused [loc (id "x"); pretty_space; atom "="; pretty_space; wrap_in_parens (expression rhs)])
    )
    ast;

  let rhs = E.assignment (Patterns.identifier "y") (E.identifier "z") in
  let ast = E.assignment (Patterns.identifier "x") rhs in
  assert_layout_of_expression
    ~ctxt
    L.(loc (fused [loc (id "x"); pretty_space; atom "="; pretty_space; expression rhs]))
    ast;

  let rhs = E.function_ () in
  let ast = E.assignment (Patterns.identifier "x") rhs in
  let expected =
    L.(loc (fused [loc (id "x"); pretty_space; atom "="; pretty_space; expression rhs]))
  in
  assert_layout_of_expression ~ctxt expected ast;
  assert_layout_of_statement ~ctxt L.(loc (fused [expected; atom ";"])) (S.expression ast);

  assert_layout_of_statement_string
    ~ctxt
    L.(
      loc
        ~loc:
          {
            Loc.none with
            Loc.start = { Loc.line = 1; column = 0 };
            _end = { Loc.line = 1; column = 8 };
          }
        (fused
           [
             wrap_in_parens
               (loc
                  ~loc:
                    {
                      Loc.none with
                      Loc.start = { Loc.line = 1; column = 1 };
                      _end = { Loc.line = 1; column = 6 };
                    }
                  (fused
                     [
                       loc
                         ~loc:
                           {
                             Loc.none with
                             Loc.start = { Loc.line = 1; column = 1 };
                             _end = { Loc.line = 1; column = 4 };
                           }
                         (group
                            [
                              atom "{";
                              indent
                                (fused
                                   [
                                     softline;
                                     loc
                                       ~loc:
                                         {
                                           Loc.none with
                                           Loc.start = { Loc.line = 1; column = 2 };
                                           _end = { Loc.line = 1; column = 3 };
                                         }
                                       (id
                                          ~loc:
                                            {
                                              Loc.none with
                                              Loc.start = { Loc.line = 1; column = 2 };
                                              _end = { Loc.line = 1; column = 3 };
                                            }
                                          "a"
                                       );
                                   ]
                                );
                              softline;
                              atom "}";
                            ]
                         );
                       pretty_space;
                       atom "=";
                       pretty_space;
                       loc
                         ~loc:
                           {
                             Loc.none with
                             Loc.start = { Loc.line = 1; column = 5 };
                             _end = { Loc.line = 1; column = 6 };
                           }
                         (id
                            ~loc:
                              {
                                Loc.none with
                                Loc.start = { Loc.line = 1; column = 5 };
                                _end = { Loc.line = 1; column = 6 };
                              }
                            "b"
                         );
                     ]
                  )
               );
             atom ";";
           ]
        )
    )
    "({a}=b);"

File: rust_port/crates/flow_parser_utils_output/src/js_layout_generator_test/assignment_precedence_test.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use flow_parser::loc::Loc;
use flow_parser::loc::Position;
use flow_parser_utils::ast_builder::expressions as E;
use flow_parser_utils::ast_builder::statements as S;
use flow_parser_utils::ast_builder::*;

use crate::layout_test_utils::layout_builder as L;
use crate::layout_test_utils::*;

#[test]
fn test() {
    let rhs = E::sequence(
        None,
        None,
        vec![
            E::identifier(None, None, "y"),
            E::identifier(None, None, "z"),
        ],
    );
    let ast = E::assignment(
        None,
        None,
        patterns::identifier(None, None, "x"),
        None,
        rhs.clone(),
    );
    assert_layout_of_expression(
        None,
        None,
        L::loc(
            None,
            L::fused(vec![
                L::loc(None, L::id(None, "x")),
                L::pretty_space(),
                L::atom("="),
                L::pretty_space(),
                L::wrap_in_parens(L::expression(None, None, &rhs)),
            ]),
        ),
        &ast,
    );

    let rhs = E::assignment(
        None,
        None,
        patterns::identifier(None, None, "y"),
        None,
        E::identifier(None, None, "z"),
    );
    let ast = E::assignment(
        None,
        None,
        patterns::identifier(None, None, "x"),
        None,
        rhs.clone(),
    );
    assert_layout_of_expression(
        None,
        None,
        L::loc(
            None,
            L::fused(vec![
                L::loc(None, L::id(None, "x")),
                L::pretty_space(),
                L::atom("="),
                L::pretty_space(),
                L::expression(None, None, &rhs),
            ]),
        ),
        &ast,
    );

    let rhs = E::function_(None, None, None, None, None, None);
    let ast = E::assignment(
        None,
        None,
        patterns::identifier(None, None, "x"),
        None,
        rhs.clone(),
    );
    let expected = L::loc(
        None,
        L::fused(vec![
            L::loc(None, L::id(None, "x")),
            L::pretty_space(),
            L::atom("="),
            L::pretty_space(),
            L::expression(None, None, &rhs),
        ]),
    );
    assert_layout_of_expression(None, None, expected.clone(), &ast);
    assert_layout_of_statement(
        None,
        L::loc(None, L::fused(vec![expected, L::atom(";")])),
        &S::expression(None, None, None, ast),
    );

    assert_layout_of_statement_string(
        L::loc(
            Some(Loc {
                source: None,
                start: Position { line: 1, column: 0 },
                end: Position { line: 1, column: 8 },
            }),
            L::fused(vec![
                L::wrap_in_parens(L::loc(
                    Some(Loc {
                        source: None,
                        start: Position { line: 1, column: 1 },
                        end: Position { line: 1, column: 6 },
                    }),
                    L::fused(vec![
                        L::loc(
                            Some(Loc {
                                source: None,
                                start: Position { line: 1, column: 1 },
                                end: Position { line: 1, column: 4 },
                            }),
                            L::group(vec![
                                L::atom("{"),
                                L::indent(L::fused(vec![
                                    L::softline(),
                                    L::loc(
                                        Some(Loc {
                                            source: None,
                                            start: Position { line: 1, column: 2 },
                                            end: Position { line: 1, column: 3 },
                                        }),
                                        L::id(
                                            Some(Loc {
                                                source: None,
                                                start: Position { line: 1, column: 2 },
                                                end: Position { line: 1, column: 3 },
                                            }),
                                            "a",
                                        ),
                                    ),
                                ])),
                                L::softline(),
                                L::atom("}"),
                            ]),
                        ),
                        L::pretty_space(),
                        L::atom("="),
                        L::pretty_space(),
                        L::loc(
                            Some(Loc {
                                source: None,
                                start: Position { line: 1, column: 5 },
                                end: Position { line: 1, column: 6 },
                            }),
                            L::id(
                                Some(Loc {
                                    source: None,
                                    start: Position { line: 1, column: 5 },
                                    end: Position { line: 1, column: 6 },
                                }),
                                "b",
                            ),
                        ),
                    ]),
                )),
                L::atom(";"),
            ]),
        ),
        "({a}=b);",
    );
}
